src/Init/Lean/Delaborator.lean

```diff
@@ -0,0 +1,371 @@
+/-
+Copyright (c) 2020 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+-/
+
+/-!
+The delaborator is the first stage of the pretty printer, and the inverse of the
+elaborator: it turns fully elaborated `Expr` core terms back into surface-level
+`Syntax`, omitting some implicit information again and using higher-level syntax
+abstractions like notations where possible. The exact behavior can be customized
+using pretty printer options; activating `pp.all` should guarantee that the
+delaborator is injective and that re-elaborating the resulting `Syntax`
+round-trips.
+
+Pretty printer options can be given not only for the whole term, but also
+specific subterms. This is used both when automatically refining pp options
+until round-trip and when interactively selecting pp options for a subterm (both
+TBD). The association of options to subterms is done by assigning a unique,
+synthetic Nat position to each subterm derived from its position in the full
+term. This position is added to the corresponding Syntax object so that
+elaboration errors and interactions with the pretty printer output can be traced
+back to the subterm.
+
+The delaborator is extensible via the `[delab]` attribute.
+-/
+
+prelude
+import Init.Lean.KeyedDeclsAttribute
+import Init.Lean.Parser.Level  -- for level quotations
+import Init.Lean.Elab
+
+namespace Lean
+
+-- TODO: move, maybe
+namespace Level
+protected partial def quote : Level → Syntax
+| zero _       => Unhygienic.run `(level|0)
+| l@(succ _ _) => match l.toNat with
+  | some n => Unhygienic.run `(level|$(mkStxNumLitAux n):numLit)
+  | none   => Unhygienic.run `(level|$(quote l.getLevelOffset) + $(mkStxNumLitAux l.getOffset):numLit)
+| max l1 l2 _  => match_syntax quote l2 with
+  | `(level|max $ls*) => Unhygienic.run `(level|max $(quote l1) $ls*)
+  | l2                => Unhygienic.run `(level|max $(quote l1) $l2)
+| imax l1 l2 _ => match_syntax quote l2 with
+  | `(level|imax $ls*) => Unhygienic.run `(level|imax $(quote l1) $ls*)
+  | l2                 => Unhygienic.run `(level|imax $(quote l1) $l2)
+| param n _    => Unhygienic.run `(level|$(mkIdent n):ident)
+-- HACK: approximation
+| mvar  n _    => Unhygienic.run `(level|_)
+
+instance HasQuote : HasQuote Level := ⟨Level.quote⟩
+end Level
+
+def getPPBinderTypes (o : Options) : Bool := o.get `pp.binder_types true
+def getPPExplicit (o : Options) : Bool := o.get `pp.explicit false
+def getPPUniverses (o : Options) : Bool := o.get `pp.universes false
+def getPPAll (o : Options) : Bool := o.get `pp.all false
+
+@[init] def ppOptions : IO Unit := do
+registerOption `pp.explicit { defValue := false, group := "pp", descr := "(pretty printer) display implicit arguments" };
```

Comment:
@leodemoura Ah, I forgot to flag the most significant change: I decided to replace `pp.implicit` with this option because I can **never** remember whether it shows or hides implicit parameters (because all other options print what they are called when *enabled*). Do you agree?

Reply:
I agree. I think this is a good change.